enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,398 @@
+       * Mitigation (Phase 2): limited automatic intervention for some use cases: for example Node1 will gain quorum only if Node2 can be verified to be down by successfully querying its BMC status.
+5. Kubelet Failure
+   1. Pacemaker’s monitoring detects the failure
+   2. Pacemaker restarts kubelet
+   3. Stop failure is optionally escalated to a node failure (fencing)
+   4. Start failure defaults to leaving the service offline
+6. Etcd Failure
+   1. Pacemaker’s monitoring detects the failure
+   2. Pacemaker either demotes etcd so it can resync, or restarts and promotes etcd
+   3. Stop failure is optionally escalated to a node failure (fencing)
+   4. Start failure defaults to leaving the service offline
+
+
+### Risks and Mitigations
+
+1. Risk: If etcd were to be made active on both peers during a network split, divergent datasets would be created
+   1. Mitigation: RHEL-HA requires fencing of a presumed dead peer before restarting etcd as a cluster-of-one
+   1. Mitigation: Peers remain inert (unable to fence peers, or start cri-o, kubelet, or etcd) after rebooting until they can contact their peer
+
+1. Risk: Multiple entities (RHEL-HA, CEO) attempting to manage etcd membership would cause an internal split-brain
+   1. Mitigation: The CEO will run in a mode that does manage not etcd membership
+
+1. Risk: Rebooting the surviving peer would require human intervention before the cluster starts, increasing downtime and creating an admin burden at remote sites
+   1. Mitigation: Lifecycle events, such as upgrades and applying new `MachineConfig`s, are not permitted in a single-node degraded state
+   1. Mitigation: Usage of the MCO Admin Defined Node Disruption [feature](https://github.com/openshift/enhancements/pull/1525) will futher reduce the need for reboots.
+   1. Mitigation: The node will be reachable via SSH and the confirmation can be scripted
+   1. Mitigation: It may be possible to identify scenarios where, for a known hardware topology, it is safe to allow the node to proceed automatically.
+
+1. Risk: “Something changed, lets reboot” is somewhat baked into OCP’s DNA and has the potential to be problematic when nodes are actively watching for their peer to disappear, and have an obligation to promptly act on that disappearance by power cycling them.
+   1. Mitigation: Identify causes of reboots, and either avoid them or ensure they are not treated as failures.
+   This may require an additional enhancement.
+
+1. Risk: We may not succeed in identifying all the reasons a node will reboot
+   1. Mitigation: ... testing? ...
+
+1. Risk: This new platform will have a unique installation flow
+   1. Mitigation: ... CI ...
+
+
+
+
+### Drawbacks
+
+The two-node architecture represents yet another distinct install type for users to choose from.
+
+The existence of 1, 2, and 3+ node control-plane sizes will likely generate customer demand to move between them as their needs change.
```

Comment:
For now, wherever your API exists for configuring etcd, should become a one way switch, and prevented from being undone

Reply:
As I understand it, this would be a change to the etcd operator where it would not allow itself to transition out of this `managedEtcdKind`: `external`.

Outside of that, the Infrastructure's `TopologyMode` shouldn't allow transitions. What manages the Infrastructure CR today? To my knowledge, we don't support transitions from SingleReplica to anything else. We can follow suit for this.

Reply:
On the other hand, PM is very keen on this _not_ being a one way switch.
There are reasonable cases to be made on both sides.

Reply:
Seems fair

Reply:
I've updated the text to better indicate that topology migration isn't supported.

Reply:
I tested a topology transition from SingleNode to HighlyAvailable and can confirm that some operators, such as ingress, are written such that they assume the topology does not change, and so they don't react correctly if the topology is changed. That said, there is nothing actually preventing a user from changing the topologies as they exist today. Since we don't actually support this (and it won't work without some development) I believe it may also be favorable to introduce a validatingadmissionpolicy to protect topology in the infrastructure config as well.

Reply:
I discussed this on the control plane arch call today, we _can_ add a validation to make this immutable, but first must make sure the field is always populated. I need to discuss with @deads2k about how this has been handled in the past

Either way, I think we should assume that transitioning is not supported and document that here

Reply:
Agreed. I feel like this should be made immutable to protect users. That said, I also want to understand how to override that so can hack on it if this becomes a requirement down the line. :smile_cat: